Add support for HTML 'title' attribute (tooltips) on Catalyst

jesxilin opened the pull request:
#### 249d9e49489f4edd4945a7bd2dd08f7fa7f7bfd7
<pre>
Add support for HTML &apos;title&apos; attribute (tooltips) on Catalyst
<a href="https://bugs.webkit.org/show_bug.cgi?id=296470">https://bugs.webkit.org/show_bug.cgi?id=296470</a>
<a href="https://rdar.apple.com/79416560">rdar://79416560</a>

Reviewed by Tim Horton.

Added support for tooltips on Catalyst.

This includes:
- Linking UIKitMacHelper
- Declaring needed protocols and interfaces
- Implementing toolTipChanged

The bulk of the functionality is in -[WKContentView _toolTipChanged:],
which creates a UIToolTipInteraction if it doesn&apos;t exist. If a
UIToolTipInteraction exists, we update the UIToolTipInteraction with
the tooltip&apos;s new string.

Notice, _toolTipChanged&apos;s has a windowChangedKeyState call and uses a
UINSSharedApplicationDelegate. This is for updating the tooltip
after the initial creation. Implementing UIToolTipInteractionDelegate&apos;s
toolTipInteraction:configurationAtPoint: is not enough to update the tooltip
because...
1. UIView&apos;s addInteraction and removeInteraction are only for separate views.
That is, if there are tooltips A and B on the same view, whichever tooltip is
triggered first will stay no matter where you hover within the view. The interaction is not
removed until you move your mouse to a different view.
2. There is no API to update (text or show/hide) the tooltip while it is displayed.

To work around #2, we need to indirectly trigger AppKit&apos;s _displayToolTipIfNecessaryIgnoringTime,
which is the logic to update the tooltip, but through UIKit.

See below for more details.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:

These files above have necessary declarations that are used for working around #2,
adding a HAVE macro, and linking UIKitMacHelper.

* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::toolTipChanged):

Calls -[WKContentView _toolTipChanged:], which contains the logic
for supporting tooltips.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _toolTipChanged:]):

This is where the support is implemented. To circumvent #2, we first use
UINSSharedApplicationDelegate, which is a UINSApplicationDelegate, then obtaining
a UINSWindow. UINSWindow has a UINSSceneView that has a NSViewDynamicToolTipManager
property. The NSViewDynamicToolTipManager can be used to call windowChangedKeyState.
We call windowChangedKeyState to trigger a chain of function calls that will eventually
trigger the update tooltip function. The chain is as follows: windowChangedKeyState calls
_restartMovementTracking which calls _displayToolTipIfNecessaryIgnoringTime, which updates
the tooltip.

(-[WKContentView toolTipInteraction:configurationAtPoint:]):

Canonical link: <a href="https://commits.webkit.org/297878@main">https://commits.webkit.org/297878@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cef410937ac28c8ba532d602ab367b631b2b4821

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113195 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32956 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23408 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119403 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63845 "Built successfully") | ✅ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33583 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41493 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86162 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41168 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ✅ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116142 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26826 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101829 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66482 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26100 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19957 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63157 "Built successfully") | | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/105712 "Found 1 new JSC stress test failure: stress/catch-from-function-inlining-simd-function.js.ftl-eager-no-cjit (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96211 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20031 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122620 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/111811 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40273 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30048 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95009 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40658 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98043 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94751 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39898 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17707 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/36409 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18200 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40159 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45658 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/136041 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39800 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36507 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/jitless-wasm-to-js-should-follow-to-wasm-call-frame-convention-2.js.default (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43133 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41537 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -302,6 +302,10 @@
 #endif
 
+#if PLATFORM(MACCATALYST)
+#define HAVE_UITOOLTIPINTERACTION 1
```
personally I would like more underscores but this is fine too

hortont424 commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -11424,6 +11424,26 @@ - (void)dropInteraction:(UIDropInteraction *)interaction sessionDidEnd:(id<UIDro
 #endif // ENABLE(DRAG_SUPPORT)
 
+#if HAVE(UITOOLTIPINTERACTION)
+- (void)_toolTipChanged:(NSString *)newToolTip {
```
{ on the next line

hortont424 commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -11424,6 +11424,26 @@ - (void)dropInteraction:(UIDropInteraction *)interaction sessionDidEnd:(id<UIDro
+
+// MARK: UIToolTipInteractionDelegate
+
+- (UIToolTipConfiguration *)toolTipInteraction:(UIToolTipInteraction *)interaction configurationAtPoint:(CGPoint)point {
```
ditto

hortont424 commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -11424,6 +11424,26 @@ - (void)dropInteraction:(UIDropInteraction *)interaction sessionDidEnd:(id<UIDro
+    }
+
+    [_toolTip setDefaultToolTip:newToolTip];
+    RetainPtr sceneView = [[UINSSharedApplicationDelegate() hostWindowForUIWindow:[self window]] sceneView];
```
Do we need to worry about the "I don't have an app" case like API tests/WKTR?

hortont424 replied:
I guess it will just all be nils and do nothing which is fine

hortont424 commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -11424,6 +11424,26 @@ - (void)dropInteraction:(UIDropInteraction *)interaction sessionDidEnd:(id<UIDro
+
+    [_toolTip setDefaultToolTip:newToolTip];
+    RetainPtr sceneView = [[UINSSharedApplicationDelegate() hostWindowForUIWindow:[self window]] sceneView];
+    [[sceneView _dynamicToolTipManager] windowChangedKeyState];
```
Let's add a FIXME that references a bug about replacing this with something less bizarre?

jesxilin committed:
Add support for HTML 'title' attribute (tooltips) on Catalyst
https://bugs.webkit.org/show_bug.cgi?id=296470
rdar://79416560

Reviewed by Tim Horton.

Added support for tooltips on Catalyst.

This includes:
- Linking UIKitMacHelper
- Declaring needed protocols and interfaces
- Implementing toolTipChanged

The bulk of the functionality is in -[WKContentView _toolTipChanged:],
which creates a UIToolTipInteraction if it doesn't exist. If a
UIToolTipInteraction exists, we update the UIToolTipInteraction with
the tooltip's new string.

Notice, _toolTipChanged's has a windowChangedKeyState call and uses a
UINSSharedApplicationDelegate. This is for updating the tooltip
after the initial creation. Implementing UIToolTipInteractionDelegate's
toolTipInteraction:configurationAtPoint: is not enough to update the tooltip
because...
1. UIView's addInteraction and removeInteraction are only for separate views.
That is, if there are tooltips A and B on the same view, whichever tooltip is
triggered first will stay no matter where you hover within the view. The interaction is not
removed until you move your mouse to a different view.
2. There is no API to update (text or show/hide) the tooltip while it is displayed.

To work around #2, we need to indirectly trigger AppKit's _displayToolTipIfNecessaryIgnoringTime,
which is the logic to update the tooltip, but through UIKit.

See below for more details.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:

These files above have necessary declarations that are used for working around #2,
adding a HAVE macro, and linking UIKitMacHelper.

* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::toolTipChanged):

Calls -[WKContentView _toolTipChanged:], which contains the logic
for supporting tooltips.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _toolTipChanged:]):

This is where the support is implemented. To circumvent #2, we first use
UINSSharedApplicationDelegate, which is a UINSApplicationDelegate, then obtaining
a UINSWindow. UINSWindow has a UINSSceneView that has a NSViewDynamicToolTipManager
property. The NSViewDynamicToolTipManager can be used to call windowChangedKeyState.
We call windowChangedKeyState to trigger a chain of function calls that will eventually
trigger the update tooltip function. The chain is as follows: windowChangedKeyState calls
_restartMovementTracking which calls _displayToolTipIfNecessaryIgnoringTime, which updates
the tooltip.

(-[WKContentView toolTipInteraction:configurationAtPoint:]):

Canonical link: https://commits.webkit.org/297878@main
